Yfac antialiasing

MichaelCotton opened the pull request:
Proposed edits:
- Added global post-processing-filter entry (and DigitalFilter object) to specify low-pass anti-aliasing filter on IQ data
- Restructured YFactorCalibration object to specify time-domain y-factor calibration
- Fixed some general text

MichaelCotton committed:
Need to include digital post processing filter

MichaelCotton committed:
Add post_processing_filter to Globals

filter object was also defined. need to tweak definitions.

MichaelCotton committed:
Changed object name from Filter to DigitalFilter

MichaelCotton committed:
Type

MichaelCotton committed:
First attempt to update YFactorCalibration object

MichaelCotton committed:
Restructured YFactorCalibration object

- Removed over-use of "additional" in global objects
- Replaced "requires" with "contains" in object definitions
- Added text to post-processing-filter description. 
- Removed single-frequency-fft object from YFactorCalibration object (from previous commit in this branch)
(I can't seem to figure out why there is no line after frequency object in YFactorCalibration object tabl)
- Added descriptive text to YFactorCalibration object
- Removed cal_source_object_enr entry from RFPath object
- Added `frequency` array to YFactorCalibration object
- Added `excess-noise-ratio` array to YFactorCalibration object 
- Fixed CR issue in Index section

bradeales commented on README.md:
```diff
@@ -143,10 +145,21 @@ Each `RFPath` object requires the following additional name/value pairs:
+
+|name|required|type|unit|description|
+|----|--------------|-------|-------|-----------|
+|`type`|false|string|N/A|E.g., `"FIR"`|
```
"type" needs a better description.

djanderson replied:
Type here refers to json type, right? So one of number, string, object, or array?

bradeales replied:
I believe the intention of `type` is to contain a description of the digital filter. The "description" field above should read something like: 'Summary description of digital filter. E.g. "FIR": Finite impulse response'

bradeales commented on README.md:
```diff
@@ -143,10 +145,21 @@ Each `RFPath` object requires the following additional name/value pairs:
 
-### 4.1.2 ScheduleEntry Object
-The `ScheduleEntry` object requires the following name/value pairs:
+### 4.1.2 Digital Filter Object
```
"Digital Filter" should be "DigitalFilter"

bradeales commented on README.md:
```diff
@@ -75,15 +76,16 @@ Per SigMF, the global object consists of name/value pairs that provide informati
 |name|required|type|unit|description|
 |----|--------------|-------|-------|-----------|
 |`sensor_definition`|false|object|N/A|Describes the sensor model components. See [SensorDefinition Object](#411-sensordefinition-object) definition. This object is RECOMMENDED.|
+|`post_processing_filter`|false|object|N/A|Describes post-processing low-pass filter appropriate for IQ captures, e.g., anti-aliasing filter. See [DigitalFilter Object](#412-digital-filter-object) definition.|
```
The name: `post_processing_filter` doesn't match the object name: "DigitalFilter Object", as we have done elsewhere. Is there a plan to use the "DigitalFilter Object" elsewhere, e.g. you might have a pre_processing_filter?

MichaelCotton replied:
@bradeales DigitalFilter will likely be used elsewhere, e.g., as we get more sophisticated with our measurement_type algorithms (performed at the sensor). I named it post_processing_filter bc we are applying a digital filter after acquisition for the ML waveforms. I see how that is ambiguous. The current motivation for this global metadata element is to specify an anti-aliasing filter, so I will be specific and name it that.

bradeales reviewed (commented):
Mainly just a question about the Digital Filter Object, and whether it will be used multiple times.

djanderson commented on README.md:
```diff
@@ -221,27 +234,31 @@ Swept-tuned detection is a standard spectrum analyzer measurement. The `SweptTun
-    "noise_levels": []
+    "receiver_setting": 2,
+    "gain": [6.1, 6.2, 6.0, 6.1, 6.4, 6.1, 6.0, 6.3, 6.2, 6.0],
+    "noise_figure": [9.1, 9.2, 9.0, 9.1, 9.4, 9.1, 9.0, 9.3, 9.2, 9.0]
```
I'd suggest keeping plural "gains" and "noise_figures" here since in all other cases where the associated value type is an array, we've used plural keys.

jawepman reviewed (commented):
Please see suggestions in text.

jawepman commented on README.md:
```diff
@@ -143,10 +145,21 @@ Each `RFPath` object requires the following additional name/value pairs:
+|`type`|false|string|N/A|E.g., `"FIR"`|
+|`length`|false|integer|N/A|Number of taps.|
+|`frequency_cutoff`|false|float|Hz|Point in the frequency response where a transition band and passband meet.|
+|`ripple_passband`|false|float|dB|Ripple in passband.|
```
Suggest adding an 'attenuation_cutoff' parameter to define the filter BW along with the frequency_cutoff.

Change frequency_cutoff description to 'Frequency at which the magnitude response decreases (from its maximum) by attenuation_cutoff.'

jawepman commented on README.md:
```diff
@@ -221,27 +234,31 @@ Swept-tuned detection is a standard spectrum analyzer measurement. The `SweptTun
+|`frequency`|false|array|Hz|Frequencies that y-factor calibrations are performed.|
+|`excess_noise_ratio`|false|array|dB|Excess noise ratio of calibrated noise source at `frequency` of y-factor calibration.|
+|`receiver_setting_name`|false|string|N/A|Name of receiver setting. Use if y-factor calibration is performed over a range of a receiver parameter, e.g., `"attenuation"`, `"input range"`.|
+|`receiver_setting_units`|false|string|N/A|Units corresponding to `receiver_setting_name`, e.g., `"dB"`, `"dBm"`.|
```
Suggest changing receiver_setting_name description to 'Name of receiver setting that affects noise figure, e.g., "attenuation", "input range". Use if receiver has an adjustable receiver setting that affects receiver noise figure (e.g., "attenuation", "input range").'

jawepman commented on README.md:
````diff
@@ -221,27 +234,31 @@ Swept-tuned detection is a standard spectrum analyzer measurement. The `SweptTun
-Example of `calibrations`:
+Example of `calibrations` for two receiver settings and ten frequencies:
 
 ```
````
Suggest adding further description of example of calibrations to make it more realistic. Something like:

Example of `calibrations` for two receiver settings and ten frequencies:
For case where receiver_setting_name is "input range" and receiver settings are -30 dBm and -20 dBm:
receiver setting: -30 dBm
gain ...
noise_figure ...

receiver setting: -20 dBm
gain: ...
noise_figure: ...

MichaelCotton committed:
Response to review comments

- Changed post_processing_filter to anti_aliasing_filter
- Improved descriptions in DigitalFilter object
- Added attenuation_cutoff to DigitalFilter object
- Changed array names in YFactorCalibration object to plural 
- Improved descriptions in YFactorCalibration object
- Made YFactorCalibration example more complete

MichaelCotton committed:
Response2 to review

Reduce number of frequencies from 10 to 5 in YFactorCalibration example

djanderson reviewed (approved):
This looks a lot better... much more readable.

bradeales commented on README.md:
```diff
@@ -221,37 +235,49 @@ Swept-tuned detection is a standard spectrum analyzer measurement. The `SweptTun
-|`reference`|false|string|N/A|Data reference point, e.g., `"receiver input"`, `"antenna output"`, `"output of isotropic antenna"`.|
+|`last_time_performed`|true|datetime|[ISO-8601](https://github.com/gnuradio/SigMF/blob/master/sigmf-spec.md#the-datetime-pair)|Date and time that calibration was performed.|
+|`frequencies`|false|array|Hz|Frequencies that y-factor calibrations are performed.|
+|`excess_noise_ratios`|false|array|dB|Excess noise ratio of calibrated noise source at `frequency` of y-factor calibration.|
```
Somewhat ignorant here ... do the `frequencies` and `excess_noise_ratios` correspond to specific `calibrations`? If so, we need to establish that relationship though a multidimensional array, and should not be based on order.

MichaelCotton replied:
I don't understand the need for a multidimensional array to establish the relationship. The y-factor cal is performed at a number of `frequencies` . At each frequency, there is an excess noise ratio (enr) that is read off the calibrated noise source - the enr is needed to calculate the gain and noise figure at the frequency. enr can probably be excluded from metadata bc there is no reall need for it after gain and noise figure are calculated. It is important, however, to specify `frequencies`.

djanderson replied:
I read it as `frequencies` is common across all calibrations, so basically the (common) x-axis labels for a number of possible gain/NF graphs... I think this is good as is.

bradeales reviewed (changes requested):
Maybe a non-sensical comment. Otherwise looks good to me.

MichaelCotton committed:
Typos - frequency --> frequencies

MichaelCotton committed:
Clean-up YFactorCalibration example

- Remove ...
